File: MCTS.py
import math
import random

from Game import Game
#from mcts_test.ticTacToeGame import TicTacToeGame
#from mcts_test.ticTacToeGame import TicTacToeGame
from policies import Tictactoe_random_rollout_policy, Tictactoe_random_expantion_policy, Tictactoe_Simulator, Go_random_expantion_policy, Go_random_rollout_policy, Go_Simulator, Go_NN_expansion_policy, Go_NN_rollout_policy
from GoGame_Logic import GoGame_Logic
class Node():
    "A node for Monte Carlo Tree Search"
    playerNum = 0 #player num in this turn (player who makes the next move, didn't make this move)
    parent = None #parent of node
    children = [] #children of this node
    wins = 0.0 #number of games won from this position
    plays = 0.0 #total number of games played from this position
    gameState = None #board in the state of after the move this node represents was played
    move = [] #move from parent that lead to this node
    C = math.sqrt(0.2) #exploration constant

    def __init__(self, _playerNum, _gameState, _move, _parent, win = 0, played = 0):
        """
        @:type _gameState Game
        :param _playerNum:
        :param _gameState:
        :param _move:
        :param _parent:
        :param win:
        :param played:
        """
        self.children = []
        self.playerNum = _playerNum
        self.gameState = _gameState.copy()
        if len(_move) != 0:
            self.gameState.makeMove((_playerNum%2)+1, _move)

        self.move = [x for x in _move]
        self.parent = _parent
        self.wins = win
        self.plays = played


    def UCT(self):
        "returns the UCT value of this node for the selection stage"
        self.wins = float(self.wins)
        return (self.wins / self.plays) + (self.C * math.sqrt(math.log(self.parent.plays, math.e) / self.plays))

    def rollOut(self, simulator, rollout_policy):
        """plays a roll-out to the end of the game & back-propagates result to all parents
        @:type rollout_policy Rollout_policy
        """

        win = simulator.simulate(self.gameState, self.playerNum, rollout_policy)

        # back-propagate rollout to all parents statistics
        temp = self

        while(temp != None):
           # print "backprop to temp:" , temp.win , temp.played, temp.playerNum
            temp.plays += 1
            #if(temp.parent != None and win != 0 and temp.parent.playerNum == win):
            if (win != 0 and (temp.playerNum%2)+1 == win):
               # print "incremented win"
                temp.wins += 1
            if(win == 3):
                temp.wins += 0.5
            temp = temp.parent

    def set_children_moves(self, moves):
        self.children_moves = moves


class MCTS():
    def __init__(self, expansion_policy, rollout_policy, simulator):
        self.expansion_policy = expansion_policy
        self.rollout_policy = rollout_policy
        self.simulator = simulator

    def get_move(self, game_state, player, rollout_times):
        TIMES = rollout_times
        root = Node(player, game_state, (), None)
        root.set_children_moves(self.expansion_policy.get_moves(root.gameState))

        # to MCTS repeatedly 'TIMES' times
        for i in range(TIMES):
            print i
            self.mcts(root)

        maxScore = -100
        maxChild = None
        #  print "total playes:" , root.plays
        #print "num of childre:" , len(root.children)
        for child in root.children:
           # print child.UCT() ,child.wins, child.plays, child.move
            #child.gameState.printMe()

            # note: I changed that in the actual move selection (not travel down the tree) will chooce that move with the highest win rate (without taking the exploration factor into account)
            child_score = child.wins / float(child.plays)  # child.UCT()
            # if child_score == 1.0:
            #     self.simulator.simulate(child.gameState, (child.playerNum%2)+1, self.rollout_policy, True)
            if (child_score > maxScore):
                maxScore = child_score
                maxChild = child

        return maxChild.move

    def get_next_expansion_move(self, node):
        "returns [] if node has all children for all possible moves or [r,c] for possible move left without child"
        # todo: work with tuples and sets(use hash on tuples) to minimize search time
        #print "LEN", len(self.expansion_policy.get_moves(node.gameState))
       # print node.gameState.board
       #  for move in self.expansion_policy.get_moves(node.gameState): #node.gameState.getAllAvailableMoves():
       #      if move not in [(child.move[0],child.move[1]) for child in node.children]:
       #          return move
       #  return []

        for move in node.children_moves: #node.gameState.getAllAvailableMoves():
            if move not in [(child.move[0],child.move[1]) for child in node.children]:
                return move
        return []

    def mcts(self, node):
        "recursive monte carlo tree search for given node"

        if (node.gameState.game_finished()):  # if node is end game leaf
            node.rollOut(self.simulator, self.rollout_policy)
            return

        possibleMoveWithoutChild = self.get_next_expansion_move(node)
        # print possibleMoveWithoutChild

        if (len(possibleMoveWithoutChild) == 0):  # if this node has a child for every possible move from its position
            # SELECTION
            # go to child with max UCT value
            childToUct = {c: c.UCT() for c in node.children}  # dictionary child to its UCT score

            # get child with max UCT score
            maxScore = max(childToUct.values())
            maxChild = None
            for child in childToUct:

                if (maxScore == childToUct[child]):
                    maxChild = child
                    break
            self.mcts(maxChild)
        else:
            # EXPANTION

            newChild = Node((node.playerNum % 2) + 1, node.gameState, possibleMoveWithoutChild, node)
            newChild.set_children_moves(self.expansion_policy.get_moves(newChild.gameState))
            newChild.rollOut(self.simulator, self.rollout_policy)  # SIMULATION & BACK PROPAGATION
            #print "adding children to size:" , len(node.children)
            node.children = node.children + [newChild]

def pvcMCTS():
    #game = TicTacToeGame()
    game = GoGame_Logic(9)
   # mcts = MCTS(Tictactoe_random_expantion_policy(100), Tictactoe_random_rollout_policy(), Tictactoe_Simulator())#MCTS(NN_Expansion_Policy(), NN_Rollout_policy())
   #  mcts = MCTS(Go_random_expantion_policy(100), Go_random_rollout_policy(),
   #             Go_Simulator())  # MCTS(NN_Expansion_Policy(), NN_Rollout_policy())
    mcts = MCTS(Go_NN_expansion_policy(5, 'nn_weights.txt'), Go_NN_rollout_policy('nn_weights.txt'),
                Go_Simulator())  # MCTS(NN_Expansion_Policy(), NN_Rollout_policy())

    while (1):
        game.printMe()
        row, col = input("Enter <row>,<col>")
        game.makeMove(1, [row, col])
        if (game.getWin() != 0):
            break
        game.printMe()
        move = mcts.get_move(game, 2, 20)

        game.makeMove(2, [move[0], move[1]])
        if (game.getWin() != 0):
            break

    game.printMe()
    if (game.getWin() != 3):
        print "player", ('X' if game.getWin() == 1 else 'Y'), "won!"
    else:
        print "tie"

def MCTS_test_against_random():
    wins= 0
    lost = 0
    tie = 0
    games = 20
    #mcts = MCTS(Tictactoe_random_expantion_policy(100), Tictactoe_random_rollout_policy(), Tictactoe_Simulator())#MCTS(NN_Expansion_Policy(), NN_Rollout_policy())
    mcts = MCTS(Go_random_expantion_policy(100), Go_random_rollout_policy(), Go_Simulator())  # MCTS(NN_Expansion_Policy(), NN_Rollout_policy())
    for x in xrange(games):
        #game = TicTacToeGame()
        game = GoGame_Logic(5)
        while (1):
            game.printMe()
           # row, col = input("Enter <row>,<col>")
            move = random.choice(game.getAllAvailableMoves())
            game.makeMove(1, [move[0], move[1]])
            if (game.getWin() != 0):
                break
           # game.printMe()
            move = mcts.get_move(game, 2, 100)

            game.makeMove(2, [move[0], move[1]])
            if (game.getWin() != 0):
                break

      #  game.printMe()
        win_state = game.getWin()
        if win_state == 1:
           # game.printMe()
            lost += 1
            print "lost", lost
        elif win_state == 2:
            wins += 1
            print "won", wins
        else:
            print "tie"
            tie += 1

    print wins, lost, tie

#MCTS_test_against_random()
#pvcMCTS()

File: MCTS/MCTS.java
package MCTS;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import MCTS.Policies.GoNNExpantionPolicy;
import MCTS.Policies.GoNNRolloutPolicy;
import goTrain.GoLogic;

public class MCTS {

	
	private GoNNExpantionPolicy expansionPolicy;
	private GoNNRolloutPolicy rolloutPolicy;
	private GoSimulator simulator;
	
	public MCTS(GoNNExpantionPolicy expansionPolicy, GoNNRolloutPolicy rolloutPolicy, GoSimulator simulator){
		this.expansionPolicy = expansionPolicy;
		this.rolloutPolicy = rolloutPolicy;
		this.simulator = simulator;
	}
	
	public int[] getMove(GoLogic gameState, int player, int rolloutTimes){
		final int TIMES = rolloutTimes;
		Node root = new Node(player,gameState, new int[0],null,0,0);
		root.setChildrenMoves(this.expansionPolicy.getMoves(root.gameState));
		
		// do MCTS repeatedly 'TIMES'
		for(int i = 0; i < TIMES; i++){
			System.out.println(i);
			this.mcts(root);
		}
		
		double maxScore = -1;
		Node maxChild = null;
		for(Node child : root.children){
			double childScore = child.wins / child.plays;
			if(child.isCaptureOrAtariMove())
				childScore = (childScore+child.move[2]) * 1.5;
			System.out.println(childScore + "," + child.wins + "," + child.plays + Arrays.toString(child.move));
			//note: I changed that in the actual root's move selection (not travel down the tree) will choce that move with the highest win rate (without taking the exploration factor into account)
			if(childScore > maxScore){
				maxScore = childScore;
				maxChild = child;
			}
		}
		return maxChild.move;
	}
	
	public int[] getNextExpansionMove(Node node){
		List<Move> alreadyChildrenOfNode = new ArrayList<Move>();
		for(Node child : node.children){
			alreadyChildrenOfNode.add(new Move(child.move[0], child.move[1]));
		}
		for(int[] move : node.childrenMoves){
			if(!alreadyChildrenOfNode.contains(new Move(move[0],move[1]))){
				return move;
			}
		}
		return new int[0];
	
		
	}
	
	/**
	 * recursive monte carlo tree search for given node
	 * @param node
	 */
	public void mcts(Node node){
		if(node.gameState.gameFinished()){
			node.rollOut(this.simulator, this.rolloutPolicy);
			return;
		}
		int[] possibleMoveWithoutChild = this.getNextExpansionMove(node);
		
		if(possibleMoveWithoutChild.length == 0){
			//selection
			//go to child with max UCT value
			double maxScore = 0;
			Node maxChild = null;
			for(Node child : node.children){
				if(child.UCT() > maxScore){
					maxScore = child.UCT();
					maxChild = child;
				}
			}
			this.mcts(maxChild);
		}else{
			//expansion
			Node newChild = new Node((node.playerNum%2)+1, node.gameState, possibleMoveWithoutChild, node,0,0);
			newChild.setChildrenMoves(this.expansionPolicy.getMoves(newChild.gameState));
			newChild.rollOut(this.simulator, this.rolloutPolicy);
			node.children.add(newChild);
		}
	}
	
	
	private static class Node{
		private static double C = Math.sqrt(2);
		private int playerNum;
		private Node parent;
		private double wins, plays;
		private GoLogic gameState;
		private int[] move;
		private List<Node> children;
		private List<int[]> childrenMoves;
		
		public Node(int _playerNum, GoLogic _gameState, int[] _move, Node _parent, double _wins, double _plays){
			
			this.children = new ArrayList<Node>();
			this.playerNum = _playerNum;
			this.gameState = _gameState.copy();
			if(_move.length != 0)
				this.gameState.makeMove((_playerNum%2)+1, _move);
			if(_move.length == 0)
				this.move = new int[0];
			else{
				if(_move.length == 3)
					this.move = new int[]{_move[0],_move[1], _move[2]};
				else
					this.move = new int[]{_move[0],_move[1]};
			}
			this.parent = _parent;
			this.wins = _wins;
			this.plays = _plays;
			
		}

		/**
		 * returns the UCT score of this node for the selection stage
		 * @return
		 */
		public double UCT(){
			return (this.wins/this.plays) + (C * Math.sqrt((Math.log(this.parent.plays)/Math.log(Math.E))/this.plays));
		}
		
		public void rollOut(GoSimulator simulator, GoNNRolloutPolicy rolloutPolicy){
			int win = simulator.simulate(this.gameState, this.playerNum, rolloutPolicy);
			Node temp = this;
			while(temp != null){
				temp.plays++;
				if(win != 0 && (temp.playerNum%2)+1 == win){
					temp.wins++;
				}
				if(win == 3)
					temp.wins += 0.5;
				
				temp = temp.parent;
			}
			
		}
		
		public void setChildrenMoves(List<int[]> moves){
			this.childrenMoves = moves;
		}
		
		public boolean isCaptureOrAtariMove(){
			return this.move.length == 3;
		}
		
	}
	
	public static class Move{
		public int row;
		public int col;
		public Move(int row, int col){
			this.row = row;
			this.col = col;
		}
		
		@Override
		public boolean equals(Object other){
			if(!(other instanceof Move))
				return false;
			Move otherMove = (Move) other;
			return row == otherMove.row && col == otherMove.col;
		}
		
		@Override
		public int hashCode(){
			int result = row;
		    result = 31 * result + col;
		    return result;
		}
		
		public String toString(){
			return (row + "," + col);
		}
		
	}
	
	
	
}

